src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala

```diff
@@ -0,0 +1,138 @@
+package com.microsoft.partnercatalyst.fortis.spark.sinks.cassandra
+
+import java.util.UUID
+
+import com.datastax.spark.connector.writer.{RowWriterFactory, WriteConf}
+import com.microsoft.partnercatalyst.fortis.spark.dto.FortisEvent
+import org.apache.spark.sql.{Dataset, SaveMode, SparkSession}
+import org.apache.spark.streaming.dstream.DStream
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.mapper.ColumnMapper
+import com.microsoft.partnercatalyst.fortis.spark.sinks.cassandra.aggregators._
+import com.microsoft.partnercatalyst.fortis.spark.sinks.cassandra.dto._
+import org.apache.spark.rdd.RDD
+import com.microsoft.partnercatalyst.fortis.spark.sinks.cassandra.udfs._
+import org.apache.spark.streaming.Time
+
+import scala.util.{Failure, Success, Try}
+
+object CassandraEventsSink{
+  private val KeyspaceName = "fortis"
+  private val TableEvent = "events"
+  private val TableEventTopics = "eventtopics"
+  private val TableEventPlaces = "eventplaces"
+  private val TableEventBatches = "eventbatches"
+  private val CassandraFormat = "org.apache.spark.sql.cassandra"
+  private val CassandraRetryIntervalMS = 2000
+  private val CassandraMaxRetryAttempts = 3
+
+  def apply(dstream: DStream[FortisEvent], sparkSession: SparkSession): Unit = {
+      val batchStream = dstream.foreachRDD{ (eventsRDD, time: Time) => {
+        if(!eventsRDD.isEmpty) {
+          val batchid = UUID.randomUUID().toString
+          val fortisEvents = writeFortisEvents(eventsRDD, batchid)
+          val aggregators = Seq(new PopularPlacesAggregator, new PopularTopicAggregator).par
+          val session = SparkSession.builder().config(eventsRDD.sparkContext.getConf)
+            .appName(eventsRDD.sparkContext.appName)
+            .getOrCreate()
+
+          if(session != null){
+            registerUDFs(session)
+            val eventBatchDF = fetchEventBatch(batchid, fortisEvents, session)
+            writeEventBatchToEventTagTables(eventBatchDF, session)
+            aggregators.map(aggregator => {
```

Comment:
Hm, still not sure we really need the in-Spark aggregation as opposed to just updating Cassandra with rows.

How much data do we expect to have in any specific batch?

If it's less than 65k, we can use Cassandra's [batch keyword](http://docs.datastax.com/en/cql/3.1/cql/cql_reference/batch_r.html) to simplify the insertion logic but just building up a bunch of simple insert statements for the RDD and then inserting them in one go.

If we can't afford one insert statement per 65k records, we can use the [copy statement](http://docs.datastax.com/en/cql/3.1/cql/cql_reference/copy_r.html) and stream the RDD rows to Cassandra.

Wondering what your thoughts are for these.